util/tracecpu/profile.go

```diff
@@ -0,0 +1,401 @@
+	task := sp.newProfileTask()
+	if err := pprof.StartCPUProfile(task.buf); err != nil {
+		// Sleep a while before retry.
+		time.Sleep(time.Millisecond)
+		return
+	}
+	ns := int(time.Second)*interval - time.Now().Nanosecond()
+	time.Sleep(time.Nanosecond * time.Duration(ns))
+	pprof.StopCPUProfile()
+	sp.sendProfileTask(task)
+}
+
+func (sp *sqlStatsProfiler) sendProfileTask(task *profileTask) {
+	task.end = time.Now().Unix()
+	sp.taskCh <- task
+}
+
+func (sp *sqlStatsProfiler) startAnalyzeProfileWorker() {
+	defer util.Recover("top-sql", "analyzeProfileWorker", nil, false)
+	for {
+		task := <-sp.taskCh
+		reader := bytes.NewReader(task.buf.Bytes())
+		p, err := profile.Parse(reader)
+		if err != nil {
+			logutil.BgLogger().Error("parse profile error", zap.Error(err))
+			continue
+		}
+		stats := sp.parseCPUProfileTags(p)
+		sp.handleExportProfileTask(p)
+		if sp.collector != nil {
+			sp.collector.Collect(task.end, stats)
+		}
+		sp.putTaskToBuffer(task)
+	}
+}
+
+type profileTask struct {
+	buf *bytes.Buffer
+	end int64
+}
+
+func (sp *sqlStatsProfiler) newProfileTask() *profileTask {
+	var task *profileTask
+	select {
+	case task = <-sp.cacheBufCh:
+		task.buf.Reset()
+	default:
+		task = &profileTask{
+			buf: bytes.NewBuffer(make([]byte, 0, 100*1024)),
+		}
+	}
+	return task
+}
+
+func (sp *sqlStatsProfiler) putTaskToBuffer(task *profileTask) {
+	select {
+	case sp.cacheBufCh <- task:
+	default:
+	}
+}
+
+func (sp *sqlStatsProfiler) parseCPUProfileTags(p *profile.Profile) []SQLStats {
+	sqlMap := make(map[string]*sqlStats)
+	idx := len(p.SampleType) - 1
+	for _, s := range p.Sample {
+		digests, ok := s.Label[labelSQLDigest]
+		if !ok || len(digests) == 0 {
+			continue
+		}
+		for _, digest := range digests {
+			stmt, ok := sqlMap[digest]
+			if !ok {
+				stmt = &sqlStats{
+					plans:      make(map[string]int64),
+					total:      0,
+					isInternal: false,
+				}
+				sqlMap[digest] = stmt
+			}
+			stmt.total += s.Value[idx]
+
+			plans := s.Label[labelPlanDigest]
+			for _, plan := range plans {
+				stmt.plans[plan] += s.Value[idx]
+			}
+		}
+	}
+	return sp.createSQLStats(sqlMap)
+}
+
+func (sp *sqlStatsProfiler) createSQLStats(sqlMap map[string]*sqlStats) []SQLStats {
+	stats := make([]SQLStats, 0, len(sqlMap))
+	for sqlDigest, stmt := range sqlMap {
+		stmt.tune()
+		for planDigest, val := range stmt.plans {
+			stats = append(stats, SQLStats{
+				SQLDigest:  sqlDigest,
+				PlanDigest: planDigest,
+				CPUTimeMs:  uint32(time.Duration(val).Milliseconds()),
+			})
+		}
+	}
+	return stats
+}
+
+type sqlStats struct {
+	plans      map[string]int64
+	total      int64
+	isInternal bool
+}
+
+// tune use to adjust stats
+func (s *sqlStats) tune() {
+	if len(s.plans) == 0 {
+		s.plans[""] = s.total
+		return
+	}
+	planTotal := int64(0)
+	for _, v := range s.plans {
+		planTotal += v
+	}
+	remain := s.total - planTotal
+	if remain <= 0 {
+		return
+	}
+	for k, v := range s.plans {
+		s.plans[k] = v + (v/planTotal)*remain
```

Comment:
Could you add some comments for this formula?

Reply:
done.